spec/java/util/Optional.lsl

```diff
@@ -0,0 +1,407 @@
+    }
+
+
+    fun ifPresent (@Generic("? super T") consumer: Consumer): void
+    {
+        required value == null || (value != null && consumer != null);
+
+        if (value != null)
+        {
+            if (consumer == null)
+                self._throwNPE();
+
+            action CALL(consumer, [value]);
+        }
+    }
+
+
+    fun ifPresentOrElse (@Generic("? super T") consumer: Consumer, emptyAction: Runnable): void
+    {
+        required value == null || (value != null && consumer != null);
+        required value != null || (value == null && emptyAction != null);
+
+        if (value != null)
+        {
+            if (consumer == null)
+                self._throwNPE();
+
+            action CALL(consumer, [value]);
+        }
+        else
+        {
+            if (emptyAction == null)
+                self._throwNPE();
+
+            action CALL(emptyAction, []);
+        }
+    }
+
+
+    fun isEmpty (): boolean
+    {
+        result = value == null;
+    }
+
+
+    fun isPresent (): boolean
+    {
+        result = value != null;
+    }
+
+
+    @Generic("U")
+    @GenericResult("U")
+    fun map (@Generic("? super T, ? extends U") mapper: Function): Optional
+    {
+        required mapper != null;
+
+        if (mapper == null)
+            self._throwNPE();
+
+        if (value == null)
+        {
+            result = _makeEmpty();
+        }
+        else
+        {
+            val mappedValue = action CALL(mapper, [value]);
+
+            if (mappedValue == null)
+                result = _makeEmpty();
+            else
+                result = new Optional(value=mappedValue);
+        }
+    }
+
+
+    @GenericResult("T")
+    fun or (@Generic("? extends Optional<? extends T>") supplier: Supplier): Optional
+    {
+        required supplier != null;
+
+        if (supplier == null)
+            self._throwNPE();
+
+        if (value == null)
+        {
+            result = action CALL(supplier, []);
+
+            if (result == null)
+                self._throwNPE();
+        }
+        else
+        {
+            result = self;
+        }
+    }
+
+
+    fun orElse (other: T): T
+    {
+        if (value == null)
+            result = other;
+        else
+            result = value;
+    }
+
+
+    fun orElseGet (@Generic("? extends T") supplier: Supplier): T
+    {
+        required supplier != null;
+
+        if (supplier == null)
+            self._throwNPE();
+
+        if (value == null)
+            result = action CALL(supplier, []);
+        else
+            result = value;
+    }
+
+
+    fun orElseThrow (): T
+    {
+        if (value == null)
+            action THROW_NEW("java.util.NoSuchElementException", ["No value present"]);
+
+        result = value;
+    }
+
+
+    @Generic("X extends Throwable")
+    @throws(["X"], generic=true)
+    fun orElseThrow (@Generic("? extends X") exceptionSupplier: Supplier): T
+    {
+        required exceptionSupplier != null;
```

Comment:
I'm not expecting the value to be present here (i.e. `value != null`) because of the fact of using this specific method indicates that the end-user is aware of the possibility of the value to be empty and ready to handle it by generating specific exception types from the supplied `Supplier` lambda.